codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/testutil/EventStreamTestTools.kt

```diff
@@ -0,0 +1,174 @@
+        codegenContext: C,
+        project: TestEventStreamProject,
+        protocol: Protocol,
+    ): RuntimeType
+
+    /** Render a builder for the given shape */
+    fun renderBuilderForShape(
+        writer: RustWriter,
+        codegenContext: C,
+        shape: StructureShape,
+    )
+}
+
+object EventStreamTestTools {
+    fun <C : CodegenContext> runTestCase(
+        testCase: EventStreamTestModels.TestCase,
+        requirements: EventStreamTestRequirements<C>,
+        codegenTarget: CodegenTarget,
+        variety: EventStreamTestVariety,
+    ) {
+        val model = EventStreamNormalizer.transform(OperationNormalizer.transform(testCase.model))
+        val serviceShape = model.expectShape(ShapeId.from("test#TestService")) as ServiceShape
+        val codegenContext = requirements.createCodegenContext(
+            model,
+            serviceShape,
+            ShapeId.from(testCase.protocolShapeId),
+            codegenTarget,
+        )
+        val test = generateTestProject(requirements, codegenContext, codegenTarget)
+        val protocol = testCase.protocolBuilder(codegenContext)
+        val generator = requirements.renderGenerator(codegenContext, test, protocol)
+
+        test.project.lib {
+            when (variety) {
+                EventStreamTestVariety.Marshall -> writeMarshallTestCases(testCase, generator)
+                EventStreamTestVariety.Unmarshall -> writeUnmarshallTestCases(testCase, codegenTarget, generator)
+            }
+        }
+        test.project.compileAndTest()
+    }
+
+    private fun <C : CodegenContext> generateTestProject(
+        requirements: EventStreamTestRequirements<C>,
+        codegenContext: C,
+        codegenTarget: CodegenTarget,
+    ): TestEventStreamProject {
+        val model = codegenContext.model
+        val symbolProvider = codegenContext.symbolProvider
+        val operationShape = model.expectShape(ShapeId.from("test#TestStreamOp")) as OperationShape
+        val unionShape = model.expectShape(ShapeId.from("test#TestStream")) as UnionShape
+
+        val project = TestWorkspace.testProject(symbolProvider)
+        val operationSymbol = symbolProvider.toSymbol(operationShape)
+        project.withModule(ErrorsModule) {
+            val errors = model.shapes()
+                .filter { shape -> shape.isStructureShape && shape.hasTrait<ErrorTrait>() }
+                .map { it.asStructureShape().get() }
+                .toList()
+            when (codegenTarget) {
+                CodegenTarget.CLIENT -> CombinedErrorGenerator(model, symbolProvider, operationSymbol, errors).render(this)
+                CodegenTarget.SERVER -> ServerCombinedErrorGenerator(model, symbolProvider, operationSymbol, errors).render(this)
+            }
+            for (shape in model.shapes().filter { shape -> shape is StructureShape && shape.hasTrait<ErrorTrait>() }) {
+                StructureGenerator(model, symbolProvider, this, shape as StructureShape).render(codegenTarget)
```

Comment:
It's annoying that we have to say `as StructureShape` even after checking `is StructureShape`. Why is that? Is it b/c we're doing the check inside a closure?

Reply:
I think the Kotlin compiler isn't smart enough to recognize the `is` check here since it happened in the filter closure.

Reply:
I simplified this in https://github.com/awslabs/smithy-rs/pull/2130.

Reply:
Simplified it a little bit more in #2107 when addressing merge conflicts.